dev/tasks/crossbow.py

```diff
@@ -70,254 +69,341 @@ def credentials(self, url, username_from_url, allowed_types):
-            Commit SHA: {sha}
-            Version: {version}
+    def __str__(self):
+        tpl = dedent('''
+            Repo: {remote}@{branch}
+            Commit: {head}
         ''')
-        logging.info(msg.format(
-            remote=self.current_remote.url,
-            branch=self.current_branch.branch_name,
-            sha=self.sha,
-            version=self.version
-        ))
+        return tpl.format(
+            remote=self.remote.url,
+            branch=self.branch.branch_name,
+            head=self.head
+        )
+
+    def fetch(self):
+        self.origin.fetch()
 
     @property
-    def sha(self):
+    def head(self):
         """Currently checked out commit's sha"""
         return self.repo.head.target
 
     @property
-    def version(self):
-        """Generate version number based on version control history"""
-        # TODO(kszucs) use self.repo.describe() instead
-        return get_version(self.path)
+    def branch(self):
+        """Currently checked out branch"""
+        reference = self.repo.head.shorthand
+        return self.repo.branches[reference]
 
     @property
-    def current_remote(self):
-        remote_name = self.current_branch.upstream.remote_name
+    def remote(self):
+        """Currently checked out branch's remote counterpart"""
+        remote_name = self.branch.upstream.remote_name
         return self.repo.remotes[remote_name]
 
     @property
-    def current_branch(self):
-        reference = self.repo.head.shorthand
-        return self.repo.branches[reference]
+    def origin(self):
+        return self.repo.remotes['origin']
 
     @property
-    def description(self):
-        return '[BUILD] {} of {}@{}'.format(self.version,
-                                            self.current_remote.url,
-                                            self.current_branch.branch_name)
+    def signature(self):
+        name = next(self.repo.config.get_multivar('user.name'))
+        email = next(self.repo.config.get_multivar('user.email'))
+        return pygit2.Signature(name, email, int(time.time()))
 
+    def parse_user_repo(self):
+        m = re.match('.*\/([^\/]+)\/([^\/\.]+)(\.git)?$', self.remote.url)
+        user, repo = m.group(1), m.group(2)
+        return user, repo
 
-class Build(object):
 
-    def __init__(self, target, name, platform, template, **params):
-        assert isinstance(target, Target)
-        assert isinstance(platform, Platform)
+class Queue(Repo):
 
-        self.name = name
-        self.target = target
-        self.platform = platform
-        self.template = template
-        self.params = params
+    def __init__(self, repo_path):
+        super(Queue, self).__init__(repo_path)
+        self._updated_refs = []
+
+    def next_job_id(self, prefix):
+        """Auto increments the branch's identifier based on the prefix"""
+        pattern = re.compile(prefix + '-(\d+)')
+        matches = list(filter(None, map(pattern.match, self.repo.branches)))
```

Comment:
Is there away to get the most recently created branch, from the server, so that we don't have to look at every branch that exists every time we create a new one?

Reply:
I'm not aware of any, but I don't think we should worry about that - with libgit2 traversing is really fast. 
If it's going to be a problem, we can address it later.

Reply:
Sounds good.